File: network/pi_prime_network.swift
import Foundation

class PiPrimeNetwork {
    let socket: Socket

    init(port: Int) {
        socket = Socket()
        socket.bind(to: port)
        socket.listen()
    }

    func listen() {
        while true {
            let client = socket.accept()
            handleConnection(client: client)
        }
    }

    private func handleConnection(client: Socket) {
        let input = client.recv(16)
        let input = Int(input!.withUnsafeBytes { $0.load(as: UInt.self) })
        let output = isPrime(input!)
        client.send([output].withUnsafeBytes { $0 })
    }

    private func isPrime(_ n: Int) -> Bool {
        if n < 2 { return false }
        for i in 2...Int(sqrt(Double(n))) {
            if n % i == 0 { return false }
        }
        return true
    }
}

let network = PiPrimeNetwork(port: 8080)
network.listen()

File: network/pi_prime_network.rb
require "socket"

class PiPrimeNetwork
    def initialize(port)
        @socket = TCPServer.new("localhost", port)
    end

    def listen
        loop do
            client = @socket.accept
            handle_connection(client)
        end
    end

    private

    def handle_connection(client)
        input = client.recv(16)
        input = input.unpack("N")[0]
        output = is_prime(input)
        client.write([output].pack("C"))
    end

    def is_prime(n)
        return false if n < 2
        (2..Math.sqrt(n)).none? { |i| n % i == 0 }
    end
end

network = PiPrimeNetwork.new(8080)
network.listen
